Ignore actual members in UnusedPrivateMember

sczerwinski opened the pull request:
The general rule of thumb should be:
- if a rule is ignored for `abstract` members, it should be ignored for `expect` members,
- if a rule is ignored for **overridden** members, it should be ignored for `actual` members.

See also:
- #3388
- #3390

Related to #3415

chao2zhang commented on detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:
```diff
@@ -204,7 +205,7 @@ private class UnusedParameterVisitor(allowedNames: Regex) : UnusedMemberVisitor(
 
     override fun visitClassOrObject(klassOrObject: KtClassOrObject) {
-        if (klassOrObject.isExpect()) return
+        if (klassOrObject.isExpect() || klassOrObject.isActual()) return
```
How is this related to the rule of thumb `if a rule is ignored for overridden members, it should be ignored for actual members.`?

sczerwinski replied:
This line is actually my mistake (I went with `isExpect()` one step too far). Actual classes or objects should be visited. Just their actual members should be ignored.

I'll update the PR.

chao2zhang commented on detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:
```diff
@@ -292,7 +294,11 @@ private class UnusedPropertyVisitor(allowedNames: Regex) : UnusedMemberVisitor(a
+            .filter {
+                (it.isPrivate() || !it.hasValOrVar()) &&
+                    it.containingClassOrObject?.isExpect() == false &&
+                    it.containingClassOrObject?.isActual() == false
```
How is this change related to the rule of thumb `if a rule is ignored for overridden members, it should be ignored for actual members.`?

sczerwinski replied:
There is no such thing as abstract or overridden constructors, but there can be `expect` or `actual` constructors, and those should be treated the same way as `expect`/`actual` functions (line 259-260).

sczerwinski replied:
I added more test cases and changed this snippet. My initial condition was excluding too many cases.

chao2zhang replied:
Thanks for the detailed explanation and additional test cases.

sczerwinski commented:
@chao2zhang, Sorry for the mistake. I've fixed the PR, rebased and squashed for better readability.

Also an example of an actual constructor, that should be ignored:
**commonMain:**
```kotlin
expect class Foo(bar: Bar) {
    expect fun doSomething()
}
```
**jvmMain:**
```kotlin
actual class Foo actual constructor(private val bar: Bar) {
    actual fun doSomething() {
        bar.doSomething()
    }
}
```
**jsMain:**
```kotlin
actual class Foo actual constructor(bar: Bar) {
    actual fun doSomething() {
        // Do something with JS that doesn't require bar
    }
}
```

chao2zhang commented on detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:
```diff
@@ -292,7 +294,10 @@ private class UnusedPropertyVisitor(allowedNames: Regex) : UnusedMemberVisitor(a
         constructor.valueParameters
-            .filter { (it.isPrivate() || !it.hasValOrVar()) && it.containingClassOrObject?.isExpect() == false }
+            .filter {
+                (it.isPrivate() || (!it.hasValOrVar() && !constructor.isActual())) &&
+                    it.containingClassOrObject?.isExpect() == false
```
I am actually planning to separate UnusedPrivateParameter from UnusedPrivateMember, so the logic here should be simplified.
